src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,157 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> currentNode = tail;
+        Node<T> newNode = new Node<>(currentNode, value, null);
+        tail = newNode;
+        if (size == 0) {
+            head = newNode;
+        } else {
+            currentNode.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (size == index) {
+            add(value);
+            return;
+        }
+        checkIndexInBounds(index);
+        addBeforeIndex(value, index);
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexInBounds(index);
+        Node<T> node = getNode(index);
+        return node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexInBounds(index);
+        Node<T> node = getNode(index);
+        if (node != null) {
+            T result = node.value;
+            node.value = value;
+            return result;
+        }
+        return value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexInBounds(index);
+        Node<T> result = getNode(index);
+        unlink(result);
+        size--;
+        return result.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = head;
+        while (node != null) {
+            if (isEquals(object, node.value)) {
+                unlink(node);
+                size--;
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNode(int index) {
```

Comment:
rewrite this method so that it can't return null - if throws an exception if the index is out of bounds, if not - it returns Node, no other option. Also, it should do it efficiently - compare the index with the middle of the length of the list - and choose where to start - from the head or from the tail